### DISCUSSION TITLE
Why Pgvecto.rs over pgvector?

### DISCUSSION THREAD

**Discussion by boriscallens** [immich-app/immich#5830] [Q&A] [answered]
Since version 1.91.0, the immich project depends on pgvecto.rs. This PostgreSQL module is in beta and currently seldomly supported beyond installing it yourself. This eliminates any hosted servers from the options.
I'm assuming it was a conscious choice to forgo the widely supported pgvector because of multithreading performance reasons?

Does this mean that everybody here is OK with running a production database in Docker?
I am not really looking forward to taking on database maintenance responsibilities.
Managing backups and restores, tuning, and maintaining a (PostgreSQL) SQL server is no fun.
It basically eliminates this solution as something I am comfortable with sharing with family and friends or other non-technical users.

Am I missing an easy solution here? What are other people doing?
What is the expected future? Will pgvecto.rs become a standard that will soon be supported by many vendors?

**Comment by mertalev** [comment, marked as answer]
pgvector vs pgvecto.rs is something we discussed at length. There are a number of reasons behind the decision to use pgvecto.rs, the most important being its much better support for filtering. pgvector has limited support for filtering; you generally have to query for the nearest rows and apply filters on that result set. This means its recall can be abysmal, and it makes it very difficult to improve search in the ways we want. pgvecto.rs doesn't have these limitations and can indeed provide as many results as requested for a set of filters. This allows for much more powerful search in a seamless way, including the ability to have infinite scroll.

Other contributing factors include much faster indexing (which also means more accurate searches because we can afford to set higher quality indexing options), much faster inserts, more features (quantization, VBASE, etc.) and the fact that the pgvecto.rs team has gone above and beyond to address any issues or concerns we had.

As for deployment, we generally expect most users to run the database in a container using the provided image. We understand there's a trade-off in that pgvector is more widely available, but we accepted this in light of these considerations.

**Comment by VoVAllen** [comment]
Hi,

I'm the developer of pgvecto.rs. pgvecto.rs is on Apache 2 license, so that won't block vendors to deploy it. We'll also consider have our own hosted service in the future. May I ask what's your current choice of hosted postgres?

**Reply by boriscallens** [reply]
Thanks for reaching out. I currently run everything in Azure on my free credits and as a learning exercise.
Nothing is stopping me from using any other service though. I had a quick look around for free tier hosting options but was unsucessfull.
I currently run the database in a container, but have yet to put time into monitoring and a relyable, tiered backup + restore system.

**Comment by nebulade** [comment]
Thanks for everyone providing explanations on this topic. I am maintaining the Immich Cloudron package and we are currently stuck due to this also. Main issue is, that on Cloudron for various reasons also to have database type specific backup/restore code, the Postgres database uses a Cloudron specific setup. Overall it's not clear if we add this extensions, if the default vectors extension used by other apps running on the same system, will see any sideffects, but more importantly this seems so specific, as not even the latest version of pgvector.rs seems to be supported by Immich and the beta status was already mentioned (it looks @VoVAllen is doing a great job though!), so this is quite hard to keep maintained in our case for a large deployment base also with regards to potential data recovery cases if things go bad.

Mostly I guess it would be good to know some outlook for this from Immich side if it makes sense for Cloudron to have specific platform quirks for only this app, if we can expect the project to be compatible with latest pgvector.rs versions in the future or if we have to drop Immich support itself for our users.

**Reply by nebulade** [reply]
Actually forgot to ask a related question. Given that both extensions seem to share the same name `vectors` why isn't there an option to use the default one shipped with postgres and take the performance hit? Or are they not actually drop-in compatible for what Immich relies on?

**Reply by mertalev** [reply]
Thanks for reaching out about this! The version of pgvecto.rs will only be pinned to 0.1.11 for a few more weeks until we upgrade to 0.2. After this upgrade, Immich will be compatible with later 0.2.* releases. 

We will also do our best to make this upgrade smooth for users. But for environments where we're not a superuser, users will have to run `ALTER EXTENSION vectors UPDATE` themselves.

**Reply by nebulade** [reply]
That is good to know. I guess to avoid potential migration issues altogether we could wait for 0.2 to happen then, if the Immich API itself will remain compatible for mobile apps until then.

I know this is just an issue where the database is shared, but could we have avoided this issue if the extension would have had its own unique name to not interfere with other apps using the extension already? I am not a postgres expert at all though :)

**Reply by VoVAllen** [reply]
Name is different actually, pgvector is `vector` and pgvecto.rs is `vectors`.  When we were naming, we just find a name that ends with "rs".

**Reply by VoVAllen** [reply]
Do you mean on cloudron.io, that some application might depend on pgvector and some might depend on pgvecto.rs, thus conflicts may happen?

**Reply by mertalev** [reply]
> Actually forgot to ask a related question. Given that both extensions seem to share the same name `vectors` why isn't there an option to use the default one shipped with postgres and take the performance hit? Or are they not actually drop-in compatible for what Immich relies on?

They're similar in some senses, but they're different enough that we would ultimately need separate search and (re-)indexing implementations for both. This would add a considerable amount of complexity and would mean we have to test and develop for both environments.

**Reply by nebulade** [reply]
Ah well that is a much better situation then, sorry for overlooking the S in the naming. So then this at least avoids the potential side-effects I had in mind. This is good news, thanks for the clarification!

**Reply by mertalev** [reply]
@nebulade You'll be able to update to 0.2 after [this](https://github.com/immich-app/immich/pull/6785) PR is merged and in a release. We'll give a heads up in the release before this as well.
